### DISCUSSION TITLE
Various booleans for song/album categorization - client dev feedback?

### DISCUSSION THREAD

**Discussion by certuna** [opensubsonic/open-subsonic-api#77] [API requests]
### Type of change

API extension

### Proposal description

The API already has some booleans in the `Child`/`AlbumID3` responses: `isVideo` and `isCompilation`
The other way to do categorization is by free text, like OS does with `releaseTypes` ("ep", "live")

See the list below for some attributes that could be set by new booleans.

Are there any of these that would be particularly interesting to client devs? Are there some other values that you'd like to receive from the server?

### API details

There are a number of track and album attributes that could in principle either be set by a specific boolean or as free text:
- `isExplicit`
- `isAudiobook` (useful for clients to know without having to implictly infer this from other free text tags?)
- `isClassical` (the popular [Classical Extras](https://github.com/MetaTunes/picard-plugins/tree/metabrainz/2.0/plugins/classical_extras) plugin of MusicBrainz Picard writes an `is_classical` tag)
- `showWorkMovement` (Apple's own tag that switches the client UI from "normal" view to "classical music" view where the Work and Movement tags are displayed, as well as more prominent display of Composer - this probably serves the same function as isClassical?)
- `isMixed` (potentially useful for clients to exclude tracks from DJ mixes from auto-playlists?)
- `isLive` (similar)
- `isComplete` (for albums, to filter out those albums that do not have the full album tracklist, but only one or two songs - many users don't like seing those listed as "complete albums" in the album list)
- `isLossless`
- `hasLyrics`

**Comment by dweymouth** [comment]
For audiobook, DJmix, and live, those are all standard release types as defined by [Picard/MusicBrainz](https://musicbrainz.org/doc/Release_Group/Type), so I'd think rather than introducing new booleans, the proper way is for the server to add those to the `releaseTypes` list if appropriate, and for the client to use that as the source of info.

An `isComplete` flag could maybe make sense, but I'm not sure how the server is supposed to know that info. (I guess if Total Tracks and/or Total Discs are set, the server could use that info, but a lot of times those won't be tagged).

for Classical, I definitely think we should add `work` and `movement` as properties to the subsonic.Child type. I would think there is no need for a separate `isClassical` flag because clients could just check if `work` and `movement` tags are set (and/or if any of the genres contain "classical" as a substring, though that is messy because some genres that aren't considered classical do have classical as a substring)

The only one I think probably doesn't need much discussion is `isExplicit` (though maybe we should just name it `explicit`), and an album should be considered explicit if *any* of its tracks are explicit. I don't think there would be much objection at all to adding this one.

**Reply by certuna** [reply]
> For audiobook, DJmix, and live, those are all standard release types as defined by [Picard/MusicBrainz](https://musicbrainz.org/doc/Release_Group/Type), so I'd think rather than introducing new booleans, the proper way is for the server to add those to the releaseTypes list if appropriate, and for the client to use that as the source of info.

The MusicBrainz releaseTypes indeed work for the album level, but there's currently no way to tell the client whether individual tracks on a release are DJ mixed or live. There are thousands (probably tens of thousands?) of albums with a combination of mixed and unmixed tracks, live and studio.

> An isComplete flag could maybe make sense, but I'm not sure how the server is supposed to know that info. (I guess if Total Tracks and/or Total Discs are set, the server could use that info, but a lot of times those won't be tagged).

My thinking is, that's a problem for servers to figure out :) The clients can never figure it out, since they only get the track/disc number through the API, not the totals. So I figured this would be a relatively easy way to pass this info along.

> for Classical, I definitely think we should add work and movement as properties to the subsonic.Child type. I would think there is no need for a separate isClassical flag because clients could just check if work and movement tags are set

Yeah this is what I've used for my (experimental) implementation of Work/Movement in Navidrome (i.e. serverside), so my question is: are the clients fine with implementing relatively messy criteria like "is Work and/or Movement set?", "is there a Composer?", "does the genre list contain Classical?", or should that logic be serverside, and the clients just prefer to get a simple answer from the server: `isClassical = true`?

> (and/or if any of the genres contain "classical" as a substring, though that is messy because some genres that aren't considered classical do have classical as a substring)

...and also not language-neutral

> The only one I think probably doesn't need much discussion is isExplicit (though maybe we should just name it explicit), and an album should be considered explicit if any of its tracks are explicit. I don't think there would be much objection at all to adding this one.

Yeah I thought so as well. The other two attributes in the API are called `isVideo` and `isCompilation` (not "compilation"), so for the sake of consistency `isExplicit` would probably be preferable?

**Reply by dweymouth** [reply]
OK, agree on `isExplicit` naming then. In my head I was thinking about the `openSubsonic=true` property we added to the base response type. But matching the naming pattern for subsonic.Child properties is more important.

For releaseTypes - admittedly I was only thinking about albums. It might make sense to add those releaseType booleans you proposed to the subsonic.Child type but not albumID3, but I'd be curious to hear others' opinions.

Regarding classical, will address in follow-up comment once I gather my thoughts...

**Reply by epoupon** [reply]
(Note that the compilation tag is not the same as the release type "compilation" set by picard)

**Reply by certuna** [reply]
By the way, Apple has an `isComplete` in their API: https://developer.apple.com/documentation/musickit/album/iscomplete

**Reply by epoupon** [reply]
Server side we could fill isComplete if the tags are populated with totaltracks and totaldiscs, but as OS states all supported fields should be set, we must distinguish the incomplete and unknown cases

**Reply by certuna** [reply]
> Server side we could fill isComplete if the tags are populated with totaltracks and totaldiscs, but as OS states all supported fields should be set, we must distinguish the incomplete and unknown cases

For that matter, same goes for isExplicit (and maybe all others?). Interestingly this is actually how the MP4 `rtng` tag is defined: https://stackoverflow.com/questions/42604812/python-mutagen-mp4-file-how-to-use-rtng-content-rating-tag
0 for unknown
2 for clean
4 for explicit

**Reply by certuna** [reply]
Just saw that MusicKit also has a `hasLyrics` attribute for the song item - I'm not sure how much we want to avoid unnecessary traffic, but at least it can prevent a client from sending out lots of useless `getLyrics` queries for non-existent lyrics, and it can probably allow clients to display a little lyrics icon in their songlists.

**Reply by certuna** [reply]
And another one could be `isLossless` - not so much for Now Playing (easy to detect that), but it would allow clients to filter for it or put a little lossless icon next to the song in songlists.

**Reply by dweymouth** [reply]
Adding on to `isLossless`, `isHD` or `isHighRes`, for tracks/albums in a format that has sample rate > 48kHz OR bit depth > 16 (Personally, I am of the opinion that anything above 16 bit/44.1 kHz lossless is unnecessary, but it's a popular user request to be able to easily identify their high res music)

**Reply by dweymouth** [reply]
Or expose sample rate, bit depth, codec directly and let the clients decide. (For albums I guess it would be based of the *minimum* stat of any track?)

**Reply by Tolriq** [reply]
Lossless have a well defined meaning based on the codec being lossless or not, HD HiRes and the rest is just personal taste that depend on users and would be problematic to have a perfect fit all values.

Sample rate bit rate codecs will need to be exposed in some way at some point for the new transcoding engine.

**Reply by certuna** [reply]
`sampleRate`, `bitrate` (already in), `codec` and `channels` are all good attributes to add, but that's probably better for another topic, this was about booleans :)

> For albums I guess it would be based of the minimum stat of any track?

Maximum I'd say - there are lots of Hi-Res albums with a couple of 'regular' bonus tracks. I'd say, "if an album has any Hi-Res material, display the album as Hi-Res". Same with lossless, multi-channel, etc. Again, we end up with the question: should the server decide this (send a isHiRes=true for the album), or do you give the client all the attributes and let the client write the logic? In this case, to ensure consistency across clients I'd probably lean towards "server decides".

**Comment by dweymouth** [comment]
OK, I think classical tagging/metadata is such a can of worms it deserves its own discussion...

**Reply by certuna** [reply]
Probably yes. Admittedly, it's very hard to do right - I think it's safe to say that nobody has really managed to implement it in a fully satisfactory way in the 25+ years that metadata tags exist.

**Reply by dweymouth** [reply]
Classical discussion here https://github.com/opensubsonic/open-subsonic-api/discussions/78
